src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintAttribute.cs

```diff
@@ -0,0 +1,46 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using Microsoft.AspNet.OData.Common;
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    /// <summary>
+    /// Represents an <see cref="Attribute"/> that can be placed on a property or placed on a class to specify the derived type constraints. 
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
```

Comment:
>[AttributeUsage(AttributeTargets.Property)] [](start = 4, length = 43)

It's only for the property, we can't put it on the class? 

My assumption is like this:

```
[DerivedTypeConstraint(typeof(Human)]
[DerivedTypeConstraint(typeof(Dog)]
....
public class Creature
{
    ....
}
```

It support to multiple config, each for one. (I am also ok to use param[])

Reply:
If you start putting it on Classes, user will expect to put constraints on entity types which we don't want to do. And Params seems cleaner. 

---
In reply to: [394517334](https://github.com/OData/WebApi/pull/1977#discussion_r394517334) [](ancestors = 394517334)

Reply:
SO, we don't have the C# attribute for entity set? 
can we do like this:

And a string argument into the Constructor of "DerivedTypeConstraintAttribute", use it to set the entity set or singleton or navigation property built from this type?

[DerivedTypeConstraint(typeof(Human), "Creatures"]
[DerivedTypeConstraint(typeof(Dog)]
....
public class Creature
{
    ....
}

---
In reply to: [399119699](https://github.com/OData/WebApi/pull/1977#discussion_r399119699) [](ancestors = 399119699,394517334)